server/src/main/java/org/opensearch/index/engine/EngineConfig.java

```diff
@@ -148,13 +150,60 @@ public Supplier<RetentionLeases> retentionLeasesSupplier() {
      * Compression Level gives a trade-off between compression ratio and speed. The higher compression level results in higher compression ratio but slower compression and decompression speeds.
      * This setting is <b>not</b> realtime updateable.
      */
-    public static final Setting<Integer> INDEX_CODEC_COMPRESSION_LEVEL_SETTING = Setting.intSetting(
+
+    public static final Setting<Integer> INDEX_CODEC_COMPRESSION_LEVEL_SETTING = new Setting<>(
         "index.codec.compression_level",
-        3,
-        1,
-        6,
+        Integer.toString(3),
+        new Setting.IntegerParser(1, 6, "index.codec.compression_level", false),
         Property.IndexScope
-    );
+    ) {
+        @Override
+        public Set<SettingDependency> getSettingsDependencies(String key) {
+            return Set.of(new SettingDependency() {
+                @Override
+                public Setting<String> getSetting() {
+                    return INDEX_CODEC_SETTING;
+                }
+
+                @Override
+                public void validate(String key, Object value, Object dependency) {
+                    if (!(dependency instanceof String)) {
+                        throw new IllegalArgumentException("Codec should be of string type.");
+                    }
+                    doValidateCodecSettings((String) dependency);
+                }
+            });
+        }
+    };
+
+    private static void doValidateCodecSettings(final String codec) {
+        switch (codec) {
+            case "zstd":
+            case "zstd_no_dict":
+                break;
+            case "best_compression":
+            case "lucene_default":
+            case "default":
+                throw new IllegalArgumentException(
+                    "Compression level cannot be set for the "
+                        + codec
+                        + " codec. Compression level settings is only applicable for zstd and zstd_no_dict codecs."
+                );
+            default:
+                if (Codec.availableCodecs().contains(codec)) {
+                    Codec luceneCodec = Codec.forName(codec);
+                    if (luceneCodec instanceof CodecSettings
+                        && ((CodecSettings) luceneCodec).supports(INDEX_CODEC_COMPRESSION_LEVEL_SETTING)) {
+                        return;
+                    }
+                }
+                throw new IllegalArgumentException(
+                    "Compression level cannot be set for the "
+                        + codec
+                        + " codec. Compression level settings is only applicable for zstd and zstd_no_dict codecs."
+                );
+        }
+    }
```

Comment:
```
throw new IllegalArgumentException(
                    INDEX_CODEC_COMPRESSION_LEVEL_SETTING.getKey() + "setting is not supported by the "        + codec
                        + " codec and cannot be set. "
                );
```

Reply:
We could remove the `Compression level settings is only applicable for zstd and zstd_no_dict codecs.` since it could be supported by custom codecs (in the future)

Reply:
ack

Reply:
done